doc/tutorial/concepts/index.md

```diff
@@ -0,0 +1,214 @@
+Constraints can be formulated ad-hoc, but this tutorial only covers *concepts*.
+Concepts are a set of constraints with a given name.
+Let's assume there is a concept called `Addable` that requires the existance of a `+` operator (as previously mentioned
+the syntax for defining concepts is not covered here).
+The following snippet demonstrates how we can constrain our function template, i.e. make the template immediately
+reject any types that don't satisfy the requirement:
+
+```cpp
+template <Addable t>
+t add(t const v1, t const v2)
+{
+    return v1 + v2;
+}
+
+int main()
+{
+    return add(1, 3); // instantiates add<int>()
+}
+```
+
+The only difference is that we have replaced `typename` with `Addable`. The standard library provides a set of
+[predefined concepts](https://en.cppreference.com/w/cpp/concepts), for our example above, the std::Integral concept
+could have been used.
+
+## Syntax variants
+
+Different constraints can be applied to different template parameters and a single template parameter can be constrained
+by multiple concepts.
+Depending on the complexity of your constraint statements, three different syntaxes are available to enforce
+constraints; all of the following are equivalent.
+
+(1) The "verbose syntax", especially useful when enforcing multiple constraints:
+
+```cpp
+template <typename t1, typename t2>
+    requires std::Integral<t1> && std::Integral<t2> // && MyOtherConcept<t1>
+auto add(t1 const v1, t2 const v2)
+{
+    return v1 + v2;
+}
+```
+
+(2) The "intermediate syntax":
+```cpp
+template <std::Integral t1, std::Integral t2>
+auto add(t1 const v1, t2 const v2)
+{
+    return v1 + v2;
+}
+```
+
+(3) The "terse syntax":
+```cpp
+auto add(std::Integral auto const v1, std::Integral const v2)
+{
+    return v1 + v2;
+}
+```
+
+\attention The terse syntax in this form is not yet available in GCC7, GCC8 and GCC9.
+
+
+Syntaxes can also be combined:
+```cpp
+template <std::Integral t1, std::Integral t2>
+    // requires MyOtherConcept<t1>
+auto add(t1 const v1, t2 const v2)
+{
+    return v1 + v2;
+}
+```
+
+# Terminology
+
+  * Template arguments can be ***constrained***.
+  * A named set of constraints is a ***concept***.
+  * A type that satisfies all requirements of a concept is said to ***model*** said concept.
+  * A *concept* that is composed of another concept and additional constraints is said to ***refine*** said concept(s).
+
+Some people confuse concepts with *interfaces*.
+Both can be used as an abstraction of concrete types, but interfaces have to be inherited from. → the abstraction
+is explicit in the definition of the type.
+Concepts on the other hand "describe properties from the outside". → types don't need to be related and don't need
+to "know about the concept" to model it.
+
+Furthermore the polymorphism possible with concepts (see below) is faster, because it is resolved at compile-time while
+interface inheritance is resolved at run-time.
+
+# Overloading and specialisation
+
```

Comment:
Maybe add something along the lines.

`Since constraints limit the set of types a function or class template may be instantiated with, it takes an important part in overloading and specialisation. `

Reply:
I have added a sentence!